src/System.Windows.Forms/src/System/Windows/Forms/Layout/ArrangedElementCollection.cs

```
@@ -127,19 +127,45 @@ private static void Copy(ArrangedElementCollection sourceList, int sourceIndex,
 
         void IList.Clear() => InnerList.Clear();
 
-        bool IList.IsFixedSize => InnerList.IsFixedSize;
+        bool IList.IsFixedSize => false;
 
-        bool IList.Contains(object? value) => InnerList.Contains(value);
+        bool IList.Contains(object? value) => (value is IArrangedElement element) ? InnerList.Contains(element) : false;
 
-        public virtual bool IsReadOnly => InnerList.IsReadOnly;
+        public virtual bool IsReadOnly => false;
 
         void IList.RemoveAt(int index) => InnerList.RemoveAt(index);
 
-        void IList.Remove(object? value) => InnerList.Remove(value);
+        void IList.Remove(object? value)
+        {
+            if (value is IArrangedElement element)
+            {
+                InnerList.Remove(element);
+            }
+        }
 
-        int IList.Add(object? value) => InnerList.Add(value);
+        int IList.Add(object? value)
+        {
+            if (value is IArrangedElement element)
```

Typing this collection is a breaking change so it is probably better to throw an argument exception here and change the documentation to make it clear that the collection will only take `IArrangedElement` now. `-1` is not something you would ever get from `ArrayList.Add`.

---

One option here is to cast the inner list to `IList` so you get the same exact throw it would do (`ArgumentException`). You'll also get the count out that way.

---

I like your idea to cast to `IList`. I'll make the change in `Add` and `IndexOf` as well.

FYI: as part of #8140, we modified several `Add`, `Contains`, `IndexOf`, etc. methods that could probably have benefited from your suggestion. Here are just a few examples: [here](https://github.com/dotnet/winforms/pull/8188/files#diff-0b3e6d3de1bcf8ce19e1a6bf0f7c6c4a4de8edcde488a696471cc55fb6470abeR36-R40), [here](https://github.com/dotnet/winforms/pull/8142/files#diff-42a2342eabb487120f84d2a4d87df7c6e59b47b2f10c2529eb725f25a4e610b7R65-R66), [here](https://github.com/dotnet/winforms/pull/8153/files#diff-0a1ebad9dd3c77e05a4f6a7bcaf4c7e72a9a0521b799b6f6a6d9908621c9bcdbR198-R199) and [here](https://github.com/dotnet/winforms/pull/8188/files#diff-0b3e6d3de1bcf8ce19e1a6bf0f7c6c4a4de8edcde488a696471cc55fb6470abeR45-R49).

---

Absolutely. I'll go back to all PRS where we refactored classes to use List instead if Array List.